File: hello-world-java/src/main/java/example/HelloResponseFilterSubscriber.java
package example;

import io.micronaut.http.MutableHttpResponse;
import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

public class HelloResponseFilterSubscriber implements Subscriber<MutableHttpResponse<?>> {

    private Subscription subscription;

    @Override
    public void onSubscribe(Subscription s) {
        subscription = s;
        if (s != null) {
            subscription.request(1);
        }

        System.out.println(String.format("%s [%s] %s", Thread.currentThread().getName(), "HelloResponseFilterSubscriber",
                s==null ? "onSubscribe: subscription is null": "onSubscribe requested 1"));
    }

    @Override
    public void onNext(MutableHttpResponse<?> response) {
        System.out.println(String.format("%s [%s] onNext invoked with response: %s", Thread.currentThread().getName(),
                "HelloResponseFilterSubscriber", response));

    }

    @Override
    public void onError(Throwable t) {
        System.out.println(String.format("%s [%s] onError", Thread.currentThread().getName(), "HelloResponseFilterSubscriber"));

    }

    @Override
    public void onComplete() {
        System.out.println(String.format("%s [%s] onComplete", Thread.currentThread().getName(), "HelloResponseFilterSubscriber"));
    }
}

File: hello-world-kotlin/src/main/kotlin/example/HelloResponseFilterSubscriber.kt
package example

import io.micronaut.http.MutableHttpResponse
import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription

class HelloResponseFilterSubscriber: Subscriber<MutableHttpResponse<*>> {

    var subscription: Subscription? = null

    override fun onComplete() {
        println("${Thread.currentThread().name} [HelloResponseFilterSubscriber] onComplete")
    }

    override fun onSubscribe(s: Subscription?) {
        subscription = s
        s?.request(1)
        println("${Thread.currentThread().name} [HelloResponseFilterSubscriber] " + if (s==null) "onSubscribe: subscription is null" else "onSubscribe requested 1")
    }

    override fun onNext(response: MutableHttpResponse<*>?) {
        println("${Thread.currentThread().name} [HelloResponseFilterSubscriber] onNext invoked with response: $response")
    }

    override fun onError(t: Throwable?) {
        println("${Thread.currentThread().name} [HelloResponseFilterSubscriber] onError")
    }
}